### DISCUSSION TITLE
I want to ask some questions about DeepMD and dp-gen

### DISCUSSION THREAD

**Discussion by DM0815** [deepmodeling/deepmd-kit#2134] [Q&A]
### Summary

Dear Zeng, Thanks for your answer yesterady. I want to ask another question. I'm not sure the kinds of Fitting network(N), Maybe it's a forward propagation network rather than a backforward progration. and I saw other dissusion. The network's output 'Energy' also will derive the distance and get the gradient 'force' , but it didn't will update the old weights. Can you tell me the 'force''s function. What can we can use it to do by 'force'. ANd it's forward progration network or backforward progration? 

 By the way, I want to ask can I regard dp-gen as a semi-supervised learning?

### DeePMD-kit Version

dp --2.1.5

### TensorFlow Version

tf --2.11.0

### Details

Dear Zeng, Thanks for your answer yesterady. I want to ask another question. I'm not sure the kinds of Fitting network(N), Maybe it's a forward propagation network rather than a backforward progration. and I saw other dissusion. The network's output 'Energy' also will derive the distance and get the gradient 'force' , but it didn't will update the old weights. Can you tell me the 'force''s function. What can we can use it to do by 'force'. ANd it's forward progration network or backforward progration? 
 By the way, I want to ask can I regard dp-gen as a semi-supervised learning?

**Comment by njzjz** [comment]
> Can you tell me the 'force''s function.

The force follows Newton's laws of motion.

$$
\vec{F} = -\vec{\nabla}E
$$

>What can we can use it to do by 'force'.

I don't understand what you mean.

> ANd it's forward progration network or backforward progration?

I don't understand what you mean by the forward propagation network or the backward propagation network. The forward propagation or the backward propagation is a computational method but does not describe what the network is.

> By the way, I want to ask can I regard dp-gen as a semi-supervised learning?

We have described it in the DP-GEN paper.

> To highlight this difference, we will call the procedures in Refs. 16, 18 “concurrent learning”. By “concurrent learning”, we mean that one does not have any data to begin with, labeled or unlabeled, and the data is generated on the fly as the training proceeds. The generation of the data and the learning is an interactive process to ensure that one obtains an “optimal” dataset which is on one hand representative enough and on the other hand as small as possible. This is in contrast to “sequential learning” in which the data is generated beforehand and the training of the model is performed afterwards. It also differs from active learning in the sense that active learning starts with unlabeled data. In a way the purpose of active learning is to find the smallest dataset that needs to be labeled in an existing unlabeled dataset.

**Reply by DM0815** [reply]
Thanks for your third answers, I got it. About the first two questions, I'm sorry to confused the gradient of loss function and output'energy'.my means is that DeepMD related paper said that Fitting network is just a MLP,but I don't see the backward propagation process, and the Lossfunction will derive the weights get the gradient, because the weight is a function about energy. and I don't know the consists of gradient if there is no virval. And the Force of Lossfunction how to derive the energy(weight's function). Therfore, I want to know the consists of gradient and backforwad propagation processing.

 In addtion, about the descriptor as an input of fitting network. Can I regard the the permutation number of atom species(4NC) as the descriptor numbers. For example , in C7O2H10 of initial paper, there are 4NC descriptors. 4NC is maximumu input size. and in the subnetwork of fitting network, the input descriptor(M1*M2), how to get the output'energy'(1*M1 or just a number), I want to figure out this process.

I'm sorry to bother you. Thanks.

**Reply by njzjz** [reply]
DeePMD-kit is based on TensorFlow. For how TensorFlow implements backward propagation, you can refer to TensorFlow's white paper ([doi: 10.48550/arXiv.1603.04467](https://doi.org/10.48550/arXiv.1603.04467), Section 4.1).

> Can I regard the the permutation number of atom species(4NC) as the descriptor numbers.

What is 4NC, and what is the descriptor number?

**Reply by DM0815** [reply]
Nc is the number of atoms within the adopted cut-off radius Rc.  The size of the input data for a sub-network is at most 4Nc.we can see the descriptor's size(M1*M2, which represent atom number and it's feature,respectively. I think), Maybe we can think M1's maximum value is 4NC?  
As for the descriptor number, in my opinion,a descriptor just is a single atom subnetwork input? For example, C7O2H10's all energy is consists of (7+2+10=19) 19 subnetwork's energy. next, a subnetwork's input(descriptor) is consists of the information of choiced atoms and it's permutated atoms.(C1 as a example, it's neighbor are C2...C7 and O1.O2,and H1....H10). Therefore, are there 19 descriptors? Or maybe I'm wrong,  there is just one descriptor in the whole fitting network ?

**Reply by njzjz** [reply]
> The size of the input data for a sub-network is at most 4Nc.

No, the size of the input data for a sub-network (if you mean the NN in the DeepPot-SE descriptor) is $N_c$.

> Maybe we can think M1's maximum value is 4NC?

No, $M_1$ is the size of the output layer of the NN in the DeepPot-SE descriptor.

> in my opinion,a descriptor just is a single atom subnetwork input?

I don't think there is a strict definition for  "one" descriptor, but for this definition, there are 19 descriptors in a $\ce{C7O2H10}$ molecule.

> In addition, if the input size is descriptor's size(M1*M2), what's the shape of output'energy'

The output is a scalar (shape=1) representing the energy of an atom.

**Reply by njzjz** [reply]
By the way, I am not sure which kind of descriptor you are talking about. Since you mention $M_1 \times M_2$, I think you want to talk about the DeepPot-SE descriptor. But the paper of $\ce{C7O2H10}$ (_Commun. Comput. Phys.,2018, 23, 629-639_) does not use the DeepPot-SE descriptor.

**Reply by DM0815** [reply]
Thanks for your information. Indeed, I saw several papers about DeePMD, but I didn't notice the difference  of them. Thank you.
